Update PoS with manual bonding process (whitelist, epoch number)

zsluedem opened the pull request:
## Overview
<!-- What this PR does, and why it's needed -->

1. add `setEpochBLockNumber` in Pos contract
2. add `addBondingRequest` in Pos contract
3. add `removeBondingRequest` in Pos contract

Only the pubkey which is in the whitelist managed by the coop could bond new validators.

And coop could config the epoch number as needed.

### Notes
<!-- Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else. -->


### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

zsluedem committed:
add setEpochBlockNumber in PoS contract

zsluedem committed:
do epoch change when blockNumber is equal to epochBlockNumber

zsluedem committed:
add bonding white list in Pos contract

zsluedem committed:
bonding require white list check

zsluedem committed:
rename PoSSpec.scala to PosSpec.scala

zsluedem committed:
fix pos tests

zsluedem committed:
fix bond status api test

zsluedem committed:
edit comment

zsluedem committed:
add test for bonding whitelist

zsluedem committed:
add test for settingEpochBlock

zsluedem committed:
add todo comment

tgrospic commented on casper/src/main/resources/Pos.rhox:
```diff
@@ -111,616 +114,690 @@ in {
+                        "coopVault" : (coopMultiVault, coopMultiVaultAddr, coopVaultHd),
+                        "posVault"  : (posVaultUnf, posVaultAddr, posVault, humanControlHandle),
+                        "currentDeployerData"   : *currentDeployerData,
+                        "epochBlockNumber"      : *epochBlockNumberCh
```
Why do we need to store epochBlockNumberCh in the state?

zsluedem replied:
Why not? Anyway. I removed it.

tgrospic replied:
Maybe we can store epoch block number only in the state. In that way it will be automatically available as state export also.

tgrospic commented on casper/src/main/resources/Pos.rhox:
```diff
@@ -111,616 +114,690 @@ in {
+                              getBlockData!(*blockDataCh) |
+                              for (@epochBlockNumber <<- epochBlockNumberCh){
+                                for (@blockNumber, _ <- blockDataCh) {
+                                  if (blockNumber % $$epochLength$$ != 0 && epochBlockNumber != blockNumber) {
```
Epoch length should not be used for now. We want to only create epoch when epochBlockNumber is not Nil and block number match current block number.

If needed we can add switch between manual and auto mode.

zsluedem replied:
Ok. I asked before and you said you want both in the standup.
I could change this.

tgrospic replied:
Sorry, I probably misunderstood you.

zsluedem committed:
remove epochBlockCh in state

zsluedem committed:
remove epoch length in closeBlock

nzpr commented:
@zsluedem there is one test failing, also could you rebase and resolve conflicts?

Isaac-DeFrain commented on casper/src/main/resources/Pos.rhox:
```diff
@@ -111,616 +114,689 @@ in {
+                              getBlockData!(*blockDataCh) |
+                              for (@epochBlockNumber <<- epochBlockNumberCh){
+                                for (@blockNumber, _ <- blockDataCh) {
+                                  if (epochBlockNumber != blockNumber) {
```
Do we intend to set an `epochBlockNumber` after each epoch and only rely on that or do we effectively want two ways of controlling when the next epoch change happens? One that's adjustable, `epochBlockNumber`, and one that's fixed but repeating, `epochLength`?

tgrospic replied:
In manual bonding procedure, which we want to use initially, `epochLength` should not be used and `epochBlockNumber` `Nil` value should be used to represent infinite value meaning no epoch change will happen.

When new validator requests are waiting or when withdrawal requests are initiated, Coop will decide on which block number epoch should happen and will set `epochBlockNumber`. After epoch change is done `epochBlockNumber` should again be reset to `Nil`.

Isaac-DeFrain replied:
Got it. So we initially set it to `Nil` and as long as `epochBlockNumber` is either `Nil` or an int greater than the current block number, no epoch change happens. We can also make it so that if `epochBlockNumber` is an int less than the current block  number, an epoch change will occur if `(blockNumber - epochBlockNumber) % epochLength == 0`, i.e. `epochLength` number of blocks have been produced sine the previous epoch change.

Isaac-DeFrain reviewed (commented):
The failing PoS tests were due `epochBlockNumber` being initialized to `Nil` in the PoS contract. Then we were using `Nil` for the epoch block number so epoch changes weren't being detected properly. `epochBlockNumber` must be explicitly set to an int to create the possibility of an epoch change.

~~It makes sense to initialize `epochBlockNumber` to `epochLength` so that's what has been done.~~

Isaac-DeFrain committed:
Merge will edits into dev

Isaac-DeFrain committed:
Initialize PoS epoch block number

- expose getEpochBlockNumber utility
- add additional error handling to setEpochBlockNumber
- change return string format on non-epoch block

Isaac-DeFrain committed:
Add invalid authority tests

- test block data with new return string formt for non-epoch blocks

Isaac-DeFrain committed:
Make epoch block number nillable

Isaac-DeFrain committed:
Set epoch block number in Pos tests

Isaac-DeFrain committed:
Reset epoch block number to Nil on epoch change

- check setEpochBlockNumber arg
- add comments

Isaac-DeFrain committed:
Add set context contract

- reset block number before each test executes

Isaac-DeFrain committed:
Move test user declarations to the top

Isaac-DeFrain committed:
Merge branch 'dev' of github.com:rchain/rchain into new-bonding-process-3739

Isaac-DeFrain commented:
bors try

Isaac-DeFrain reviewed (approved):
LGTM

Isaac-DeFrain commented:
bors try
